runtime/tests/checker/assignment_test.go

```diff
@@ -124,15 +124,136 @@ func TestCheckInvalidAssignmentTargetExpression(t *testing.T) {
 
 	t.Parallel()
 
-	_, err := ParseAndCheck(t, `
-      fun f() {}
+	t.Run("function invocation result", func(t *testing.T) {
 
-      fun test() {
-          f() = 2
-      }
-    `)
+		t.Parallel()
 
-	errs := ExpectCheckerErrors(t, err, 1)
+		_, err := ParseAndCheck(t, `
+          fun f() {}
+
+          fun test() {
+              f() = 2
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+	})
+
+	t.Run("index into function invocation result", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          fun f(): [Int] {
+              return [1]
+          }
+
+          fun test() {
+              f()[0] = 2
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+	})
+
+	t.Run("assess member of function invocation result", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          struct S {
+              var x: Int
+
+              init() {
+                  self.x = 1
+              }
+          }
+
+          let s = S()
+
+          fun f(): S {
+              return s
+          }
+
+          fun test() {
+              f().x = 2
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+	})
+
+	t.Run("index into identifier", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          let xs = [1]
+
+          fun test() {
+              xs[0] = 2
+          }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("access member of identifier", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          struct S {
+              var x: Int
+
+              init() {
+                  self.x = 1
+              }
+          }
+
+          let s = S()
+
+          fun test() {
+              s.x = 2
+          }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("index into array literal", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          fun test() {
+              [1][0] = 2
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+	})
+
+	t.Run("index into dictionary literal", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          fun test() {
+              {"a": 1}["a"] = 2
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
 
-	assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+		assert.IsType(t, &sema.InvalidAssignmentTargetError{}, errs[0])
+	})
```

Comment:
Test cases look great. Do we need a test case to verify that if a function returns an internal value, then changes to the returned value won't change the internal value?

Reply:
Good idea, I'll add some

Reply:
Added a test in 69cb18b

Reply:
Looks good to me 👍